protocols/kad/src/behaviour.rs

```diff
@@ -922,21 +895,44 @@ where
     /// The result of this operation is delivered in a
     /// reported via [`KademliaEvent::OutboundQueryCompleted{QueryResult::GetProviders}`].
     pub fn get_providers(&mut self, key: record::Key) -> QueryId {
-        let providers = self
+        let providers: HashSet<_> = self
             .store
             .providers(&key)
             .into_iter()
             .filter(|p| !p.is_expired(Instant::now()))
             .map(|p| p.provider)
             .collect();
+
         let info = QueryInfo::GetProviders {
             key: key.clone(),
-            providers,
+            count: 1,
+            providers_found: providers.len(),
         };
-        let target = kbucket::Key::new(key);
+
+        let target = kbucket::Key::new(key.clone());
         let peers = self.kbuckets.closest_keys(&target);
         let inner = QueryInner::new(info);
-        self.queries.add_iter_closest(target.clone(), peers, inner)
+        let id = self.queries.add_iter_closest(target.clone(), peers, inner);
+
+        // No queries were actually done for the results yet.
+        let stats = QueryStats::empty();
+
+        self.queued_events
+            .push_back(NetworkBehaviourAction::GenerateEvent(
+                KademliaEvent::OutboundQueryProgressed {
+                    id,
+                    result: QueryResult::GetProviders(Ok(GetProvidersOk {
+                        key,
+                        providers_so_far: providers.len(),
+                        providers,
```

Comment:
Why return an event in case `providers` is empty? Same for `get_record` above.

Reply:
you are totally right, fixed this